processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQueryRunnerFactory.java

```diff
@@ -81,6 +92,33 @@ public TimeBoundaryQueryRunner(Segment segment)
       this.adapter = segment.asStorageAdapter();
     }
 
+    private Function<Cursor, Result<DateTime>> skipToFirstMatching = new Function<Cursor, Result<DateTime>>()
+    {
+      @Override
+      public Result<DateTime> apply(Cursor cursor)
+      {
+        if (cursor.isDone()) { return null; }
+        final LongColumnSelector timestampColumnSelector = cursor.makeLongColumnSelector(Column.TIME_COLUMN_NAME);
+        DateTime timestamp = new DateTime(timestampColumnSelector.get());
+        return new Result<>(adapter.getInterval().getStart(), timestamp);
+      }
+    };
+
+    private DateTime getTimeBoundary(StorageAdapter adapter, TimeBoundaryQuery legacyQuery, boolean descending)
+    {
+      final Sequence<Result<DateTime>> resultSequence = QueryRunnerHelper.makeCursorBasedQuery(
+          adapter, legacyQuery.getQuerySegmentSpec().getIntervals(),
+          Filters.toFilter(legacyQuery.getDimensionsFilter()),
+          descending, new AllGranularity(), skipToFirstMatching
+      );
+      List<Result<DateTime>> resultList = Sequences.toList(resultSequence, Lists.<Result<DateTime>>newArrayList());
```

Comment:
You don't need to materialize the entire list

Sequences.toList(Sequences.limit(resultSequence, 1), Lists.<List<Long>>newArrayList()).get(0);